src/main/kotlin/org/jitsi/nlj/transform/node/incoming/Vp9Parser.kt

```diff
@@ -0,0 +1,144 @@
+ * from frames, and also diagnoses packet format variants that the Jitsi videobridge won't be able to route.
+ */
+class Vp9Parser(
+    parentLogger: Logger
+) : ObserverNode("Vp9 parser") {
+    private val logger = createChildLogger(parentLogger)
+    // Stats
+    private var numKeyframes: Int = 0
+    private var sources: Array<MediaSourceDesc> = arrayOf()
+
+    private val pictureIdState = StateChangeLogger("missing picture id", logger)
+    private val extendedPictureIdState = StateChangeLogger("missing extended picture ID", logger)
+    private var numSpatialLayers = -1
+
+    /** Encodings we've actually seen.  Used to clear out inferred-from-signaling encoding information. */
+    private val ssrcsSeen = HashSet<Long>()
+
+    override fun observe(packetInfo: PacketInfo) {
+        val vp9Packet = packetInfo.packet as Vp9Packet
+
+        ssrcsSeen.add(vp9Packet.ssrc)
+
+        if (vp9Packet.hasScalabilityStructure) {
+            // TODO: handle case where new SS is from a packet older than the
+            //  latest SS we've seen.
+            val packetSpatialLayers = vp9Packet.scalabilityStructureNumSpatial
+            if (packetSpatialLayers != -1) {
+                if (numSpatialLayers != -1 && numSpatialLayers != vp9Packet.scalabilityStructureNumSpatial) {
+                    packetInfo.layeringChanged = true
+                }
+                numSpatialLayers = packetSpatialLayers
+            }
+            val srcAndEnc = findRtpEncodingDesc(vp9Packet)
+            if (srcAndEnc != null) {
+                val (src, enc) = srcAndEnc
+                val newEnc = vp9Packet.getScalabilityStructure(eid = enc.layers[0].eid)
+                if (newEnc != null) {
+                    src.setEncodingLayers(newEnc.layers, vp9Packet.ssrc)
+                }
+                for (otherEnc in src.rtpEncodings) {
+                    if (!ssrcsSeen.contains(otherEnc.primarySSRC)) {
+                        src.setEncodingLayers(emptyArray(), otherEnc.primarySSRC)
+                    }
+                }
+            }
+
+            /* TODO: we need a way to restore the encoding desc's old layer set if it switches back to some other codec
+             *  (i.e. VP8)
+             */
+        }
+        /* VP9 marks keyframes in every packet of the keyframe - only count the start of the frame so the count is correct. */
+        /* Alternately we could keep track of keyframes we've already seen, by timestamp, but that seems unnecessary. */
+        if (vp9Packet.isKeyframe && vp9Packet.isStartOfFrame) {
+            logger.cdebug { "Received a keyframe for ssrc ${vp9Packet.ssrc} ${vp9Packet.sequenceNumber}" }
+            numKeyframes++
+        }
+        if (vp9Packet.spatialLayerIndex > 0 && vp9Packet.isInterPicturePredicted) {
+            /* Check if this layer is using K-SVC. */
+            /* TODO: should we somehow track the keyframe layers separately in
```

Comment:
Do we need to resolve this TODO?

Reply:
I don't think we do for the initial release - as we discussed, the bitrate of keyframes is probably small enough it doesn't matter.  I can change the comment to make this clearer, though.